File: Runtime/Cell.vb
'*********************************************************************************************
' ST-Sim: A SyncroSim Module for the ST-Sim State-and-Transition Model.
'
' Copyright © 2007-2017 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.
'
'*********************************************************************************************

''' <summary>
''' The simulation cell class
''' </summary>
''' <remarks>
''' There can be a huge number of these so:
''' (1.) Don't add any data members to this class unless you absolutely need to
''' (2.) Don't make any allocations unless they are required for the model run
''' </remarks>
Public Class Cell

    Private m_CellId As Integer
    Private m_StratumId As Integer
    Private m_SecondaryStratumId As Nullable(Of Integer)
    Private m_TertiaryStratumId As Nullable(Of Integer)
    Private m_StateClassId As Integer
    Private m_Age As Integer
    Private m_TstValues As New TstCollection
    Private m_Transitions As New List(Of Transition)
    Private m_Keys As Dictionary(Of String, Object)

    ''' <summary>
    ''' Constructor
    ''' </summary>
    ''' <param name="cellID">The ID for this cell</param>
    ''' <remarks></remarks>
    Public Sub New(ByVal cellId As Integer)
        Me.m_CellId = cellId
    End Sub

    ''' <summary>
    ''' Unique integer Id for the cell
    ''' </summary>
    Public ReadOnly Property CellId As Integer
        Get
            Return Me.m_CellId
        End Get
    End Property

    ''' <summary>
    ''' Stratum Id for the cell
    ''' </summary>
    Public Property StratumId As Integer
        Get
            Return Me.m_StratumId
        End Get
        Set(ByVal value As Integer)
            Me.m_StratumId = value
        End Set
    End Property

    ''' <summary>
    ''' Gets or sets the secondary stratum Id for the cell
    ''' </summary>
    ''' <value></value>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Public Property SecondaryStratumId As Nullable(Of Integer)
        Get
            Return Me.m_SecondaryStratumId
        End Get
        Set(value As Nullable(Of Integer))
            Me.m_SecondaryStratumId = value
        End Set
    End Property

    ''' <summary>
    ''' Gets or sets the tertiary stratum Id for the cell
    ''' </summary>
    ''' <returns></returns>
    Public Property TertiaryStratumId As Nullable(Of Integer)
        Get
            Return Me.m_TertiaryStratumId
        End Get
        Set(ByVal value As Nullable(Of Integer))
            Me.m_TertiaryStratumId = value
        End Set
    End Property

    ''' <summary>
    ''' StateClass Id for the cell
    ''' </summary>
    Public Property StateClassId As Integer
        Get
            Return Me.m_StateClassId
        End Get
        Set(ByVal value As Integer)
            Me.m_StateClassId = value
        End Set
    End Property

    ''' <summary>
    ''' Age of the cell
    ''' </summary>
    ''' <remarks></remarks>
    Public Property Age As Integer
        Get
            Return Me.m_Age
        End Get
        Set(ByVal value As Integer)
            Me.m_Age = value
        End Set
    End Property

    ''' <summary>
    ''' Collection of TST values for the cell
    ''' </summary>
    ''' <value></value>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Friend ReadOnly Property TstValues As TstCollection
        Get
            Return Me.m_TstValues
        End Get
    End Property

    ''' <summary>
    ''' Collection of transitions for the cell
    ''' </summary>
    ''' <remarks></remarks>
    Friend ReadOnly Property Transitions As IList(Of Transition)
        Get
            Return Me.m_Transitions
        End Get
    End Property

    ''' <summary>
    ''' Associates an object with this cell
    ''' </summary>
    ''' <param name="key"></param>
    ''' <param name="value"></param>
    ''' <remarks></remarks>
    Public Sub SetAssociatedObject(ByVal key As String, ByVal value As Object)

        If (Me.m_Keys Is Nothing) Then
            Me.m_Keys = New Dictionary(Of String, Object)
        End If

        Me.m_Keys.Add(key, value)

    End Sub

    ''' <summary>
    ''' Gets an associated object for this cell
    ''' </summary>
    ''' <param name="key"></param>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Public Function GetAssociatedObject(ByVal key As String) As Object

        If (Me.m_Keys Is Nothing) Then
            Return Nothing
        End If

        Return Me.m_Keys(key)

    End Function

End Class

File: src/Runtime/Cell.cs
// stsim: A SyncroSim Package for developing state-and-transition simulation models using ST-Sim.
// Copyright © 2007-2024 Apex Resource Management Solutions Ltd. (ApexRMS). All rights reserved.

using System.Collections.Generic;

namespace SyncroSim.STSim
{
    /// <summary>
    /// The simulation cell class
    /// </summary>
    /// <remarks>
    /// There can be a huge number of these so:
    /// (1.) Don't add any data members to this class unless you absolutely need to
    /// (2.) Don't make any allocations unless they are required for the model run
    /// </remarks>
    public class Cell
    {
        private int m_CellId;
        private int m_CollectionIndex;
        private int m_StratumId;
        private int? m_SecondaryStratumId;
        private int? m_TertiaryStratumId;
        private int m_StateClassId;
        private int m_Age;
        private TstCollection m_TstValues = new TstCollection();
        private List<Transition> m_Transitions = new List<Transition>();
        private Dictionary<string, object> m_Keys;

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="cellId">The ID for this cell.</param>
        /// <param name="rasterIndex">The collection index for this cell.</param>
        /// <remarks></remarks>
        public Cell(int cellId, int rasterIndex)
        {
            this.m_CellId = cellId;
            this.m_CollectionIndex = rasterIndex;
        }

        /// <summary>
        /// Unique integer Id for the cell
        /// </summary>
        /// <remarks>
        /// If the model is run spatially, this ID is also the index into the raster used to generate the cell.
        /// </remarks>
        public int CellId
        {
            get
            {
                return this.m_CellId;
            }
        }

        /// <summary>
        /// Gets this cell's index in the cell collection.
        /// </summary>
        /// <remarks>
        /// This value can be used to index into rasters that have been compressed to match the cell collection.
        /// </remarks>
        public int CollectionIndex
        {
            get
            {
                return this.m_CollectionIndex;
            }
        }

        /// <summary>
        /// Stratum Id for the cell
        /// </summary>
        public int StratumId
        {
            get
            {
                return this.m_StratumId;
            }
            set
            {
                this.m_StratumId = value;
            }
        }

        /// <summary>
        /// Gets or sets the secondary stratum Id for the cell
        /// </summary>
        /// <value></value>
        /// <returns></returns>
        /// <remarks></remarks>
        public int? SecondaryStratumId
        {
            get
            {
                return this.m_SecondaryStratumId;
            }
            set
            {
                this.m_SecondaryStratumId = value;
            }
        }

        /// <summary>
        /// Gets or sets the tertiary stratum Id for the cell
        /// </summary>
        /// <returns></returns>
        public int? TertiaryStratumId
        {
            get
            {
                return this.m_TertiaryStratumId;
            }
            set
            {
                this.m_TertiaryStratumId = value;
            }
        }

        /// <summary>
        /// StateClass Id for the cell
        /// </summary>
        public int StateClassId
        {
            get
            {
                return this.m_StateClassId;
            }
            set
            {
                this.m_StateClassId = value;
            }
        }

        /// <summary>
        /// Age of the cell
        /// </summary>
        /// <remarks></remarks>
        public int Age
        {
            get
            {
                return this.m_Age;
            }
            set
            {
                this.m_Age = value;
            }
        }

        /// <summary>
        /// Collection of TST values for the cell
        /// </summary>
        /// <value></value>
        /// <returns></returns>
        /// <remarks></remarks>
        public TstCollection TstValues
        {
            get
            {
                return this.m_TstValues;
            }
        }

        /// <summary>
        /// Collection of transitions for the cell
        /// </summary>
        /// <remarks></remarks>
        public IList<Transition> Transitions
        {
            get
            {
                return this.m_Transitions;
            }
        }

        /// <summary>
        /// Associates an object with this cell
        /// </summary>
        /// <param name="key"></param>
        /// <param name="value"></param>
        /// <remarks></remarks>
        public void SetAssociatedObject(string key, object value)
        {
            if (this.m_Keys == null)
            {
                this.m_Keys = new Dictionary<string, object>();
            }

            this.m_Keys.Add(key, value);
        }

        /// <summary>
        /// Gets an associated object for this cell
        /// </summary>
        /// <param name="key"></param>
        /// <returns></returns>
        /// <remarks></remarks>
        public object GetAssociatedObject(string key)
        {
            if (this.m_Keys == null)
            {
                return null;
            }

            return this.m_Keys[key];
        }
    }
}
